server/src/main/java/org/elasticsearch/transport/InboundDecoder.java

```diff
@@ -0,0 +1,241 @@
+        if (isOnHeader()) {
+            int messageLength = TcpTransport.readMessageLength(reference);
+            if (messageLength == -1) {
+                return 0;
+            } else if (messageLength == 0) {
+                fragmentConsumer.accept(PING);
+                return 6;
+            } else {
+                int headerBytesToRead = headerBytesToRead(reference);
+                if (headerBytesToRead == 0) {
+                    return 0;
+                } else {
+                    totalNetworkSize = messageLength + TcpHeader.BYTES_REQUIRED_FOR_MESSAGE_SIZE;
+
+                    Header header = readHeader(messageLength, reference);
+                    bytesConsumed += headerBytesToRead;
+                    if (header.isCompressed()) {
+                        decompressor = new TransportDecompressor(recycler);
+                    }
+                    fragmentConsumer.accept(header);
+
+                    if (isDone()) {
+                        finishMessage(fragmentConsumer);
+                    }
+                    return headerBytesToRead;
+                }
+            }
+        } else if (isDecodingFailed()) {
+            int bytesToConsume = Math.min(reference.length(), totalNetworkSize - bytesConsumed);
+            bytesConsumed += bytesToConsume;
+            if (isDone()) {
+                finishMessage(fragmentConsumer);
+            }
+            return bytesToConsume;
+        } else {
+            // There are a minimum number of bytes required to start decompression
+            if (decompressor != null && decompressor.canDecompress(reference.length()) == false) {
+                return 0;
+            }
+            int bytesToConsume = Math.min(reference.length(), totalNetworkSize - bytesConsumed);
+            bytesConsumed += bytesToConsume;
+            ReleasableBytesReference retainedContent;
+            if (isDone()) {
+                retainedContent = reference.retainedSlice(0, bytesToConsume);
+            } else {
+                retainedContent = reference.retain();
+            }
+            if (decompressor != null) {
+                decompress(retainedContent);
+                ReleasableBytesReference decompressed;
+                while ((decompressed = decompressor.pollDecompressedPage()) != null) {
+                    consumeNonEmptyContent(decompressed, fragmentConsumer);
+                }
+            } else {
+                consumeNonEmptyContent(retainedContent, fragmentConsumer);
+            }
+            if (isDone()) {
+                finishMessage(fragmentConsumer);
+            }
+
+            return bytesToConsume;
+        }
+    }
+
+    @Override
+    public void close() {
+        isClosed = true;
+        cleanDecodeState();
+    }
+
+    private void consumeNonEmptyContent(ReleasableBytesReference content, Consumer<Object> fragmentConsumer) {
```

Comment:
Do we even need this method now that we fixed the decompression to not produce any empty content? It seems impossible to get here with an empty `content`?